// containers/good/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import * as R from 'ramda'
import {Link} from 'react-router-dom'

import {fetchGoodById, addGoodToBasket} from 'actions'
import {getGoodById} from 'selectors'
import BasketCart from 'components/basketCart'

class Good extends Component {
    componentDidMount() {
        this.props.fetchGoodById(this.props.match.params.id)
    }

    renderFields() {
        const {good} = this.props

        const columnFields = R.compose(
            R.toPairs,
            R.pick([
                'cpu',
                'camera',
                'size',
                'weight',
                'display',
                'battery',
                'memory'
            ])
        )(good)

        return columnFields.map(([key, value]) => (
            <div className='column' key={key}>
                <div className='ab-details-title'>
                    <p>{key}</p>
                </div>
                <div className='ab-details-info'>
                    {value}
                </div>
            </div>
        ))
    }

    renderContent() {
        const {good} = this.props
        return (
            <div className='thumbnail'>
                <div className='row'>
                    <div className='col-md-6'>
                        <img 
                            className='img-thumbnail'
                            src={good.image}
                            alt={good.name}
                        />
                    </div>
                    <div className='col-md-6'>
                        {this.renderFields()}
                    </div>
                </div>
                <div className='caption-full'>
                    <h4 className='pull-right'>${good.price}</h4>
                    <h4>{good.name}</h4>
                    <p>{good.description}</p>
                </div>
            </div>
        )
    }

    renderSidebar() {
        const {good, addGoodToBasket} = this.props
        return (
            <div>
                <p className='lead'>Quick shop</p>
                <BasketCart />
                <div className='form-group'>
                    <h1>{good.name}</h1>
                    <h2>${good.price}</h2>
                </div>
                <Link to='/' className='btn btn-info btn-block'>Back to Store</Link>
                <button
                    type='button'
                    className='btn btn-success btn-block'
                    onClick={() => addGoodToBasket(good.id)}
                >  
                Add to cart 
                </button>
            </div>
        )
    }

    render() {
        const {good} = this.props
        return (
            <div className='view-container'>
                <div className='container'>
                    <div className='row'>
                        <div className='col-md-9'>
                            {good && this.renderContent()}
                        </div>
                        <div className='col-md-3'>
                            {good && this.renderSidebar()}
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
         good: getGoodById(state, state.goodPage.id)
    }
}

const mapDispatchToProps = {
    fetchGoodById,
    addGoodToBasket
}

export default connect(mapStateToProps, mapDispatchToProps)(Good)
